src/python/pants/backend/visibility/rule_types.py

```diff
@@ -131,82 +170,97 @@ def create_parser_state(
             cast(BuildFileVisibilityRules, dependents_rules).get_action(
-                target_adaptor,
-                source_path,
-                relpath=target_path,
+                target=dependency_adaptor,
+                address=origin_address,
+                relpath=dependency_address.spec_path,
                 outgoing_dependency=False,
             )
             if dependents_rules is not None
-            else DependencyRuleAction.ALLOW
+            else (None, DependencyRuleAction.ALLOW, None)
         )
-        if incoming is None:
-            dep_rules = cast(BuildFileVisibilityRules, dependents_rules)
-            raise BuildFileVisibilityRulesError(
-                f"There is no matching dependents rule for the `{target_adaptor.type_alias}` "
-                f"target {target_path}:{target_adaptor.name or os.path.basename(target_path)} "
-                f"for the dependency from the `{source_adaptor.type_alias}` target {source_path}:"
-                f"{source_adaptor.name or os.path.basename(source_path)} in {dep_rules.path}"
+        if in_action is None:
+            raise BuildFileVisibilityRulesError.create(
+                kind="__dependents_rules__",
+                rules=cast(BuildFileVisibilityRules, dependents_rules),
+                ruleset=in_ruleset,
+                origin_address=origin_address,
+                origin_adaptor=origin_adaptor,
+                dependency_address=dependency_address,
+                dependency_adaptor=dependency_adaptor,
             )
-        return incoming if incoming != DependencyRuleAction.ALLOW else outgoing
+        if in_action is DependencyRuleAction.DENY or out_action is DependencyRuleAction.ALLOW:
+            action = in_action
+        else:
+            action = out_action
+        source_rule = f"{out_ruleset}[{out_pattern}]" if out_ruleset else origin_address.spec_path
+        target_rule = f"{in_ruleset}[{in_pattern}]" if in_ruleset else dependency_address.spec_path
+        return DependencyRuleApplication(
+            action=action,
+            rule_description=f"{source_rule} -> {target_rule}",
+            origin_address=origin_address,
+            origin_type=origin_adaptor.type_alias,
+            dependency_address=dependency_address,
+            dependency_type=dependency_adaptor.type_alias,
+        )
 
     def get_action(
         self,
         target: TargetAdaptor,
-        path: str,
+        address: Address,
         relpath: str,
         outgoing_dependency: bool,
-    ) -> DependencyRuleAction | None:
+    ) -> tuple[VisibilityRuleSet | None, DependencyRuleAction | None, str | None]:
         """Get applicable rule for target type from `path`.
 
         The rules are declared in `relpath`.
         """
         ruleset = self.get_ruleset(target)
         if ruleset is None:
-            return None
+            return None, None, None
+        path = str(address) if address.is_file_target else address.spec_path
```

Comment:
@danxmoran this line here enables the use of globs like `src/python/**/my_file.py` or what you'd like to have... in case the target gen is named differently than the directory, beware that you'd need to consider that in the pattern.. `src/python/**/myfile.py:sometarget` (or with a wildcard, of course)

Reply:
So `src/python/**/myfile.py*` would match both cases?